### ISSUE TITLE
[Direct3D12] Creating complicated root signature with descriptor ranges is not supported well enough.

### ISSUE THREAD

**Issue by bdachev** [sharpdx/SharpDX#609]
I tried to create a root signature in code containing descriptor range (RootParameterType.DescriptorTable) which requires to define RootDescriptorTable entry which in turn needs PDescriptorRanges of type IntPtr. In native API you separately define an array of D3D12_DESCRIPTOR_RANGE elements and reference them in D3D12_ROOT_DESCRIPTOR_TABLE elements. In .NET that would mean unsafe and fixed unless you come up with another solution.

**Comment by RobyDX** [comment]
Hi, I'm using alpha02 from nuget 
I'm doing this way:
First the ranges

```javascript
DescriptorRange range = new DescriptorRange()
{
    RangeType = DescriptorRangeType.ConstantBufferView,
    BaseShaderRegister = 0,
    DescriptorCount = 1,
    OffsetInDescriptorsFromTableStart = int.MinValue,
};
```

Only work when OffsetInDescriptorsFromTableStart is the max UInt (so equal to int.MinValue).
Then I use System.Runtime.InteropServices.Marshal to get pointer.

```javascript
RootParameter parameter = new RootParameter()
{
    ShaderVisibility = ShaderVisibility.Vertex,
    ParameterType = RootParameterType.DescriptorTable,
    DescriptorTable = new RootDescriptorTable()
    {
        DescriptorRangeCount = 1,
        PDescriptorRanges = Marshal.UnsafeAddrOfPinnedArrayElement(new DescriptorRange[] { range }, 0)
    }
};
```
Am I doing well? Code works with this
On Direct3D12 samples there are some utils for descriptor range. Can you introduce them please?

**Comment by xoofx** [comment]
I have pushed some fixes when using the `RootParameter` API. But I haven't done anykind of tests, so if you are testing it, let me know if it is working or not! Thanks!

**Comment by RobyDX** [comment]
No problem, is a pleasure. I'll test as soon as possible (I've downloaded master_integration but it doesn't compile, SharpDx code generator give a fatal error)

**Comment by xoofx** [comment]
it should compile. Have you installed [Windows SDK 10.0.10240.0](https://dev.windows.com/en-us/downloads/windows-10-sdk)?

**Comment by RobyDX** [comment]
Yes. 
The master branch compile, the master_integration not (githhub said error too). The error seems related to Direct2D. Here the build

Severity	Code	Description	Project	File	Line
Error		in gccxml C:/Program Files (x86)/Windows Kits/10/Include/10.0.10240.0/um/d2d1_3.h:921: error: repeated using declaration 'using ID2D1Device::CreateDeviceContext'	SharpDX	G:\SharpDX\Source\SharpDX\SharpGen	1

Severity	Code	Description	Project	File	Line
Error		in gccxml C:/Program Files (x86)/Windows Kits/10/Include/10.0.10240.0/um/d2d1_3.h:962: error: repeated using declaration 'using ID2D1Factory1::CreateDevice'	SharpDX	G:\SharpDX\Source\SharpDX\SharpGen	1

**Comment by xoofx** [comment]
Sorry, forgot to commit one file, that explain your error

**Comment by bdachev** [comment]
Hi xoofx,
I did some tests of the new RootParameter API as you call it. ;)
My test project seems to work as expected.
What I would suggest as further improvement to add to RootParameter constructors argument for ShaderVisibility flag. If I have to define it all in a single array now I do something like that:

var rsparams = new RootParameter[]
{
    new RootParameter(new RootDescriptor(), RootParameterType.ConstantBufferView)
    {
        ShaderVisibility = ShaderVisibility.Vertex,
    },
    new RootParameter(new []
    {
        new DescriptorRange
        {
            RangeType = DescriptorRangeType.ConstantBufferView,
            BaseShaderRegister = 1,
            DescriptorCount = 1,
        },
    })
    {
        ShaderVisibility = ShaderVisibility.Vertex,
    },
    ...
Also it might be handy to have RootParameter constructor accepting DescriptorRange[] + offset and size arguments so that part of bigger DescriptorRange array is referenced as it is possible with native API.
Hope it all makes some sense! :)
B.

**Comment by xoofx** [comment]
Yeah, @bdachev sorry forgot the ShaderVisibility parameter while refactoring yesterday, so it should be fixed by latest commit d13c03bcf843f0d9a8a7239ebd87003e27671bf5
